src/batch/src/executor/monitor/stats.rs

```diff
@@ -46,12 +44,189 @@ impl BatchMetrics {
         )
         .unwrap();
 
+        // Spawn a deletor.
+        // TaskMetricsManager will create task metrics for each BatchExecution and task metrics will
+        // record all the label it used. When the BatchExecution is finished, the task metrics will
+        // send all record to the delete queue. deletor is responsible for reading
+        // unused record from delete queue periodically and remove them from metrics.
+        // The deletor will remove the record after several minutes. Because it's important not to
+        // delete it immediately since prometheus pull data periodically.
+        let (delete_queue_sender, mut delete_queue_receiver) =
+            tokio::sync::mpsc::channel::<DeleteRecord>(1024);
+        let exchange_recv_row_number_clone = exchange_recv_row_number.clone();
+        tokio::spawn(async move {
+            let mut interval = tokio::time::interval(Duration::from_secs(60));
+            let mut delete_cache: Vec<DeleteRecord> = Vec::new();
+            let mut connect = true;
+            while connect {
+                // run every minute.
+                let _ = interval.tick().await;
+
+                // delete all record in delete_cache .
+                while let Some(delete_record) = delete_cache.pop() {
+                    let query_id = delete_record.query_id;
+                    for (metric_name, vec) in &delete_record.records {
+                        match metric_name.as_str() {
+                            "exchange_recv_row_number" => {
+                                for [source_stage_id, target_stage_id, source_task_id, target_task_id] in
+                                    vec
+                                {
+                                    if exchange_recv_row_number_clone
+                                        .remove_label_values(&[
+                                            &query_id,
+                                            &source_stage_id.to_string(),
+                                            &target_stage_id.to_string(),
+                                            &source_task_id.to_string(),
+                                            &target_task_id.to_string(),
+                                        ])
+                                        .is_err()
+                                    {
+                                        // Already remove, just skip it.
+                                    }
+                                }
+                            }
+                            _ => {
+                                unimplemented!("Never reach here");
+                            }
+                        }
+                    }
+                }
+
+                // read from delete queue and push into delete_cache.
```

Comment:
Why design a cache here? What about directly delete when read from receiver?

Reply:
When we read from receiver, we can't guarantee it have been pull by prometheus. (Think about a scene: A metrics just be create and the execution is done). So I put it in cache and then delete it next time to guarantee this metrics exist at least more than 1 min.